File: Longest Common Prefix.java
class Solution 
{
    public String longestCommonPrefix(String[] strs) 
    {
        if(strs.length == 0) return "";

        int totalString = strs.length;
        
        //Assuming the first string is the most common prefix.
        String prefix = strs[0]; 

        //'i' starts at 1 instead of 0 because there's no need to compare the first string in the array.
        for(int i=1;i<totalString;i++) 
        {
            //'toCompare' will be assigned every string in each iteration.
            String toCompare = strs[i]; 
            
            //While 'prefix' is not equal to current 'toCompare', cut the last character of 'prefix' until the condition is met.
            //As long as 'prefix' is not equal to 'toCompare', take out the last character of prefix until they are equal.
            while(!prefix.equals(toCompare) && toCompare.indexOf(prefix)!= 0)  
            {
                prefix = prefix.substring(0,prefix.length()-1);
            }
        }
        return prefix;
    }
}
// ["flower","flow","flight"]
//First off, assume the first string in the array is the longest common prefix. ------------> (In this case, flower)
//Then, compare every other strings in the array to shorten/modify (if necessary) the prefix.

//String a.indexOf(String b) method     ----->     (If string b is longer than string a, then it will always return 0 because there's no way b can go into a.) 
//flow.indexOf(flower) will return -1 because 'flower' cannot go into 'flow'.
//flower.indexOf(flow) will return 0 because 'flow' can go into 'flower', starting with index 0;                

File: Longest Common Prefix.py
https://leetcode.com/problems/longest-common-prefix/
  
Write a function to find the longest common prefix string amongst an array of strings.
If there is no common prefix, return an empty string "".

Input: strs = ["flower","flow","flight"]
Output: "fl"
  
class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        if len(strs) == 1:
            return strs[0]
        
        prefix = ''     
        index = 0
        
        """
        Build the prefix based on the first str in strs
        Take each char in the first str and compare it with the other strs to see if the char is the same.
        Use index to move the char we want to check in other strs.
        The longest substring is ALWAYS constrained by the shortest string in strs.
        """
        for char in strs[0]:
            for s in range(1, len(strs)):
                cur_str = strs[s]
                
                if index > len(cur_str)-1 or char != cur_str[index]:
                    return prefix
                
            prefix += char
            index += 1
                
        return prefix
